internals/overlord/state/identities_test.go

```diff
@@ -0,0 +1,536 @@
+		"bill": {
+			Access: "admin",
+		},
+	})
+	c.Assert(err, ErrorMatches, `identity "bill" invalid: identity must have at least one type \("local"\)`)
+
+	// Ensure user IDs are unique with existing users.
+	err = st.AddIdentities(map[string]*state.Identity{
+		"bill": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 1000},
+		},
+	})
+	c.Assert(err, ErrorMatches, `identities "bill" and "mary" cannot both have user ID 1000`)
+
+	// Ensure user IDs are unique among the ones being added.
+	err = st.AddIdentities(map[string]*state.Identity{
+		"bill": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 2000},
+		},
+		"bale": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 2000},
+		},
+	})
+	c.Assert(err, ErrorMatches, `identities "bale" and "bill" cannot both have user ID 2000`)
+}
+
+func (s *identitiesSuite) TestUpdateIdentities(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	original := map[string]*state.Identity{
+		"bob": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 42},
+		},
+		"mary": {
+			Access: state.AdminAccess,
+			Local:  &state.LocalIdentity{UserID: 1000},
+		},
+	}
+	err := st.AddIdentities(original)
+	c.Assert(err, IsNil)
+
+	err = st.UpdateIdentities(map[string]*state.Identity{
+		"bob": {
+			Access: state.AdminAccess,
+			Local:  &state.LocalIdentity{UserID: 1000},
+		},
+		"mary": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 42},
+		},
+	})
+	c.Assert(err, IsNil)
+
+	// Ensure they were updated correctly.
+	identities := st.Identities()
+	c.Assert(identities, DeepEquals, map[string]*state.Identity{
+		"bob": {
+			Name:   "bob",
+			Access: state.AdminAccess,
+			Local:  &state.LocalIdentity{UserID: 1000},
+		},
+		"mary": {
+			Name:   "mary",
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 42},
+		},
+	})
+
+	// Can't update identity names that don't exist.
+	err = st.UpdateIdentities(map[string]*state.Identity{
+		"bill": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 43},
+		},
+		"bale": {
+			Access: state.ReadAccess,
+			Local:  &state.LocalIdentity{UserID: 42},
+		},
+		"mary": {
+			Access: state.AdminAccess,
+			Local:  &state.LocalIdentity{UserID: 1000},
+		},
+	})
+	c.Assert(err, ErrorMatches, "identities do not exist: bale, bill")
+
+	// Ensure validation is being done (full testing done in AddIdentity).
+	err = st.UpdateIdentities(map[string]*state.Identity{
+		"bob": nil,
+	})
+	c.Assert(err, ErrorMatches, `identity "bob" invalid: identity must not be nil`)
+
+	// Ensure unique user ID testing is being done (full testing done in AddIdentity).
+	err = st.UpdateIdentities(map[string]*state.Identity{
+		"bob": {
+			Access: state.AdminAccess,
+			Local:  &state.LocalIdentity{UserID: 42},
+		},
+	})
+	c.Assert(err, ErrorMatches, `identities "bob" and "mary" cannot both have user ID 42`)
```

Comment:
Need a test with three identities with the same user ID.

Reply:
Added, thanks.